server/routes/search.js

```diff
@@ -1,52 +1,78 @@
-        meals = await mealsController.findAllMeals(query);
-        if (!query) {
-            res.json(chefs);
+        const { id } = req.user;
+        const { primaryAddress } = await userController.findOneWithId(id);
+        const { lat, lng } = primaryAddress || {};
+        if (req.query.searchType === "chefs") {
+            let { radiusKm } = req.query;
+            radiusKm = radiusKm && Number(radiusKm);
+            const query = req.query.cuisine
+                ? { cuisineSpecialty: { $in: req.query.cuisine.split(",").map(c => RegExp(c, 'i')) } }
+                : {};
+            const chefs = (await chefsController.findAllChefs(query))
+                .filter((chef) => chef.userId && chef.userId.primaryAddress)
+                .map((chef) => {
+                    const {
+                        userId: { primaryAddress },
+                    } = chef;
+                    const distanceKm = coordinatesDistanceCalc(
+                        lat,
+                        lng,
+                        primaryAddress.lat,
+                        primaryAddress.lng
+                    );
+                    return { ...chef._doc, distanceKm };
+                })
+                .filter((chef) => {
+                    if (!radiusKm) return true;
+                    return chef.distanceKm < radiusKm;
+                });
+
+            if (!query) {
+                res.json(chefs);
+                return;
+            }
+            const meals = await mealsController.findAllMeals({
+                chefId: { $in: chefs.map((chef) => chef._id) },
+            });
+
+            chefMeals = {};
+            chefs.forEach((chef) => (chefMeals[chef._id] = []));
+            meals.forEach((meal) => chefMeals[meal.chefId].push(meal));
+            res.json({ chefs, chefMeals });
         }
-        const chefs = await chefsController.findAllChefs(
-            { _id: { $in: meals.map((meal) => meal.chefId) } });
-        
-        const chefIndex = {}
-        chefs.forEach(chef => chefIndex[chef._id] = chef)
-        meals.forEach(meal => meal["chefId"]= chefIndex[meal.chefId])
 
-        res.json({ meals  });
+        if (req.query.searchType === "meals") {
+            const query = req.query.cuisine
+                ? { cuisineType: { $in: req.query.cuisine.split(",").map(c => RegExp(c, 'i')) } }
+                : {};
+            const meals = await mealsController.findAllMeals(query);
+            if (!query) {
+                res.json(chefs);
+            }
+            const chefs = await chefsController.findAllChefs({
+                _id: { $in: meals.map((meal) => meal.chefId) },
+            });
 
-    }
-    
-    return;
-});
+            const chefIndex = {};
+            chefs.forEach((chef) => (chefIndex[chef._id] = chef));
+            meals.forEach((meal) => (meal["chefId"] = chefIndex[meal.chefId]));
+            res.json({ meals });
+        }
+    })
+);
```

Comment:
This whole route is a bit hard to follow. I'm thinking it might be easier to read if you pull some of the work out into functions?

Reply:
I agree, I refactored it

Reply:
This is much better!